src/lib.rs

```diff
@@ -1601,6 +1602,40 @@ pub trait Itertools : Iterator {
         self.next().map(move |x| self.fold(x, f))
     }
 
+    /// Accumulate the elements in the iterator in a balanced manner.
+    ///
+    /// This produces a result like `f(f(1, 2), f(3, 4))`, a balanced call tree,
+    /// instead of `f(f(f(1, 2), 3), 4)`, a linear sequence.
+    ///
+    /// If `f` is associative, prefer the normal `fold1` instead.
+    ///
+    /// ```
+    /// use itertools::Itertools;
+    ///
+    /// assert_eq!((0..9).fold1_balanced(|x, y| x - y), Some(-8));
+    /// let num_strings = (0..5).map(|x| x.to_string());
+    /// assert_eq!(num_strings.fold1_balanced(|x, y| format!("({}-{})", x, y)),
+    ///     Some(String::from("(((0-1)-(2-3))-4)")));
+    /// assert_eq!((0..0).fold1_balanced(|x, y| x * y), None);
+    /// ```
```

Comment:
I think we can skip the first example that gives `-8` here. The 0..4 example should use 1..5 or 1..6 just like the initial example in this doc comment. Maybe it's easier to read with format string `"({} + {})"`, at least with the extra spacing.

(We try to keep in mind that examples serve as documentation.. if it needs better testing, we use tests for that. It's easy to be lazy here though, it happens. I wouldn't *mind* a quickcheck test for this method :slightly_smiling_face: )

Reply:
I realized I'd mistranslated my pop code to fold instead of rfold, and quickcheck totally found it (on the len-7 case, too, which is minimal).  Cool library!